### PR TITLE
made commonsparsefeatures more scalable...

### PR DESCRIPTION
**Pull request by tomerk**
at large numbers of machines and high feature counts.

See issue #183

### CODE REVIEW

File: src/main/scala/nodes/util/CommonSparseFeatures.scala [hunk 1]
```diff
@@ -27,6 +27,8 @@ case class CommonSparseFeatures[T : ClassTag](numFeatures: Int) extends Estimato
     }
   }
 
+  def merge(a: Array[(T, (Int, Long))], b: Array[(T, (Int, Long))]): Array[(T, (Int, Long))] = (a ++ b).sorted(ordering.reverse).take(numFeatures)
```
Comment (etrain) [inline src/main/scala/nodes/util/CommonSparseFeatures.scala]: pretty long line.

File: src/main/scala/nodes/util/CommonSparseFeatures.scala [hunk 2]
```diff
@@ -34,7 +36,7 @@ case class CommonSparseFeatures[T : ClassTag](numFeatures: Int) extends Estimato
     val featureFrequenciesWithUniqueId = featureOccurrences.reduceByKey {
       (x, y) => (x._1 + y._1, Math.min(x._2, y._2))
     }
-    val mostCommonFeatures = featureFrequenciesWithUniqueId.top(numFeatures)(ordering).map(_._1)
+    val mostCommonFeatures = featureFrequenciesWithUniqueId.glom().map(_.sorted(ordering.reverse).take(numFeatures)).treeReduce(merge).map(_._1)
```
Comment (etrain) [inline src/main/scala/nodes/util/CommonSparseFeatures.scala]: same here

Comment (shivaram) [inline src/main/scala/nodes/util/CommonSparseFeatures.scala]: also a comment here on why we are doing this would be good.

**Comment by etrain** [comment]
This might be a good test case for `OptimizableNode` can we come up with a rule for when its better to do this vs. the old way?

**Comment by shivaram** [comment]
Isn't the difference just that we do a `treeReduce` instead of a `reduce` ? If so it should be possible to just configure `treeReduce` to behave like `reduce` when the data is small.  Pasting the `top` function implementation below. 

```
      val mapRDDs = mapPartitions { items =>
        // Priority keeps the largest elements, so let's reverse the ordering.
        val queue = new BoundedPriorityQueue[T](num)(ord.reverse)
        queue ++= util.collection.Utils.takeOrdered(items, num)(ord)
        Iterator.single(queue)
      }
      mapRDDs.reduce { (queue1, queue2) =>
          queue1 ++= queue2
          queue1
      }.toArray.sorted(ord)
```

**Comment by tomerk** [comment]
Sure, I'll use that instead of the glom and sorted code.

**Comment by shivaram** [comment]
Yeah and if you use the bounded queue we shouldn't see any perf degradation in terms of compute

**Comment by tomerk** [comment]
Ah, spark's bounded priority queue is marked private. Should I copy the code or do something else?

**Comment by shivaram** [comment]
According to http://stackoverflow.com/questions/7878026/is-there-a-priorityqueue-implementation-with-fixed-capacity-and-custom-comparato there is one in Guava. Not sure what our take on using guava is. (@etrain ?) Otherwise the answer suggested there or inserted into a sorted set is not bad

**Comment by etrain** [comment]
I like the `BoundedPriorityQueue` approach. While adding a dependency on Guava for this is kind of a bummer, I'd rather that than reimplementing ourselves.

**Comment by tomerk** [comment]
@etrain `BoundedPriorityQueue` is a Spark private util, not a guava class. As for the guava one mentioned in the stack overflow post: It's not a full bounded priority queue implementation, and the comments on that answer warn that it shouldn't be used for bounded priority queues:
"We are giving some consideration to deprecating MinMaxPriorityQueue, just because it tends to get misused in this way."

**Comment by etrain** [comment]
Alright - what do we need? A thing that takes in an iterator over the partition and returns the top k elements by inserting/removing elements from a binary heap or something? I guess this is probably simple enough to implement ourselves but i don't want to go too deep down a rabbit hole here. Maybe there's another implementation somewhere we should be using?

**Comment by tomerk** [comment]
bounded priority stuff pushed.

**Comment by etrain** [comment]
Hmm - do we not have unit tests for `CommonSparseFeatures` - seems like this would be a good time to put them in to make sure that the priority queue stuff isn't breaking some edge cases. Otherwise this LGTM.

**Comment by tomerk** [comment]
We already do, as part of the SparseFeaturesSuite

File: src/main/scala/nodes/util/CommonSparseFeatures.scala [hunk 3]
```diff
@@ -34,8 +40,25 @@ case class CommonSparseFeatures[T : ClassTag](numFeatures: Int) extends Estimato
...
     val featureSpace = mostCommonFeatures.zipWithIndex.toMap
     new SparseFeatureVectorizer(featureSpace)
   }
+
+  /**
+   * Returns the first K elements from the input as defined by the specified implicit Ordering[T]
+   * and maintains the ordering.
+   */
+  def takeOrdered[T](input: Iterator[T], num: Int)(implicit ord: Ordering[T]): Seq[T] = {
+    val ordering = new com.google.common.collect.Ordering[T] {
```
Comment (shivaram) [inline src/main/scala/nodes/util/CommonSparseFeatures.scala]: I don't think we explicitly depend on guava in our build.sbt and we seem to be getting it through some transitive dependencies ?

Comment (tomerk) [inline src/main/scala/nodes/util/CommonSparseFeatures.scala]: Transitive via Spark, I'll add it as a direct dependency

**Comment by tomerk** [comment]
Okay, I added an explicit guava dependency. We may have to keep an eye on this as we update Spark versions.

**Comment by shivaram** [comment]
How did you pick the version number ?

**Comment by tomerk** [comment]
It was what all the existing transitive dependencies on Guava had resolved to.

**Comment by shivaram** [comment]
Sounds good. LGTM
